palaeorotate

LewisAJones opened the pull request:
# Review palaeorotate()

## Description

This function generates palaeogeographic coordinates for fossil occurrence data. It does so by spatiotemporally linking present-day geographic coordinates and age estimates with a spatial grid (1° x 1°) rotated to the midpoint of stratigraphic stages (Geological Timescale, 2020). Three plate rotation models are available in the function ("Merdith2021", "Scotese2018", and "Wright2013"). Furthermore, uncertainty in palaeocoordinates can also be easily calculated (range in latitude and longitude between plate rotation models).

## Motivation and Context

Generating palaeocoordinates for fossil occurrence data can be time consuming and computational intensive (depending on the size of your dataset). While some fossil occurrence databases include pre-generated palaeocoordinates for fossil occurrences, it is often unclear which model these databases use and the workflow behind rotating them. Recent ongoing work has also shown great differences between palaeogeographic reconstructions from different plate rotation models. Traditionally, palaeobiologists tend to use one plate rotation model to generate palaeocoordinates for fossil occurrences and treat it as there is no uncertainty in this. 

Closes #11

KEichenseer reviewed (commented):
### Review 1:
The function works very well, great Job Lewis! Below I suggest some changes:

**Description:**
Replace _distribution_ with _position_ when referring to a single occurrence
In “Details”, add that nearby points will be assigned equivalent palaeocoordinates due to the spatial binning

**Code:**
**1.** Just a style point: “lon” would be easier to spell / remember than “lng”, 

**2.** circumvent re-downloading the rotation data when using the function multiple times. This could be achieved by adding an argument "store_data" or similar, and if desired the rotation data could then be saved to the global environment with 
assign("palaeo_rots", readRDS(paste0(files, "/Merdith2021.RDS")), envir=globalenv()). 
This would require to rename it to “palaeo_rots_meredith” (scotese,...) and then reassigning it to palaeo_rots, depending on the model specification. Before downloading, we would check exists("palaeo_rots_meredith"), and then it wouldn’t need to download if it already exists. Alternatively, we could not delete the file in the tempdir() directory and then reload it into the session if it exists, checking with 
if(!file.exists(paste0(files, "/Merdith2021.RDS")))
This would obviate the need for renaming variables, and the rotation files would not clutter the current session and would be deleted when the R session is closed.

**3.** could we read the rotation file directly from the internet, rather than downloading it first and then reading it in? It should be possible with the url, e.g. http://mgimond.github.io/ES218/Week02b.html
This could potentially save some time, as the downloading and readRDS both take some time: 
![image](https://user-images.githubusercontent.com/40391500/180659388-761704fc-8dc2-4c2f-a691-3d1061ecba36.png)


**4.** Error handling: 
Should we allow or add an option to allow NA values for any of lat, lng and age and return NA rotated ages, instead of throwing an error? This would allow the user to easily rotate data sets with some missing values.

    Further: Add an error message in the beginning that aborts the function if either of the columns are not numeric or contain NA, e.g. 
    if (any(!is.numeric(x$lat) | is.na(x$lat)) |
        any(!is.numeric(x$lng) | is.na(x$lng)) |
        any(!is.numeric(x$age) | is.na(x$age))) {
      stop("lng, lat and age need to be numeric")
    }

    Add: Throw an error when longitude is outside of c(-180,180), analogous to the error message thrown for unreasonable latitude values

LewisAJones commented:
> ### Review 1:
> The function works very well, great job Lewis! Below I suggest some changes:

Hi @KEichenseer, thanks for all the great comments! See below for implemented changes.

> 
> **Description:** Replace _distribution_ with _position_ when referring to a single occurrence. In “Details”, add that nearby points will be assigned equivalent palaeocoordinates due to the spatial binning.

Implemented.

> 
> **Code:** **1.** Just a style point: “lon” would be easier to spell / remember than “lng”,

I don't disagree, but "lng" is used quite a lot for spatial data, including the PBDB. I've opted to retain "lng" for now because I think people are used to working with this. Although we are not specifically developing the package for PBDB users, I suspect many of our users will be using that data.

> 
> **2.** circumvent re-downloading the rotation data when using the function multiple times. This could be achieved by adding an argument `store_data `or similar, and if desired the rotation data could then be saved to the global environment with `assign("palaeo_rots", readRDS(paste0(files, "/Merdith2021.RDS")), envir=globalenv())`. This would require to rename it to _palaeo_rots_meredith_ (_scotese_,...) and then reassigning it to `palaeo_rots`, depending on the model specification. Before downloading, we would check `exists("palaeo_rots_meredith")`, and then it wouldn’t need to download if it already exists. Alternatively, we could not delete the file in the `tempdir()` directory and then reload it into the session if it exists, checking with `if(!file.exists(paste0(files, "/Merdith2021.RDS")))` This would obviate the need for renaming variables, and the rotation files would not clutter the current session and would be deleted when the R session is closed.

Great suggestion! I've now implemented the file exists check as I think this is the cleaner approach. 

> 
> **3.** could we read the rotation file directly from the internet, rather than downloading it first and then reading it in? It should be possible with the url, e.g. http://mgimond.github.io/ES218/Week02b.html This could potentially save some time, as the downloading and `readRDS() `both take some time: ![image](https://user-images.githubusercontent.com/40391500/180659388-761704fc-8dc2-4c2f-a691-3d1061ecba36.png)

With the updates of the file exists check, I think this is best left as it is. The initial run of the function will still run at the same speed, but if I read the file directly, I will still need to save it after.

> 
> **4.** Error handling: Should we allow or add an option to allow `NA` values for any of lat, lng and age and return `NA` rotated ages, instead of throwing an error? This would allow the user to easily rotate data sets with some missing values.
> 

My preference here is to flag an error message. If the data can not be palaeorotated, then it can't be used in the subsequent analysis? We provide a very explicit error message for this. I've been caught out a few times with functions not working in the past just because I had NAs in lng/lat (but received no useful error message). As a user, I think I would personally prefer to be thrown an error if some of my data is missing/NA so I can at least go check what's going on. When working with large datasets, people might miss this otherwise.

> Further: Add an error message in the beginning that aborts the function if either of the columns are not numeric or contain `NA`, e.g.
> 
> ```
> if (any(!is.numeric(x$lat) | is.na(x$lat)) |
>      any(!is.numeric(x$lng) | is.na(x$lng)) |
>      any(!is.numeric(x$age) | is.na(x$age))) {
>    stop("lng, lat and age need to be numeric")
>  }
> ```
> 
> Add: Throw an error when longitude is outside of c(-180,180), analogous to the error message thrown for unreasonable latitude values

Thanks! All those have now been added.

KEichenseer reviewed (approved):
Excellent, thanks for implementing the changes! I'm happy with the function now.

LewisAJones commented:
Hi @KEichenseer, as recently discussed I've now added additional functionality. This functionality is to allow point rotations by calling the GPlates API. When you have chance, I would appreciate any additional feedback you may have! Thanks!

Buffan3369 commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
+      # Add unique ages to API request
+      names(API_request) <- uni_ages
+
+      # Run API call with progress bar
```
This is super fancy, I like it!

Buffan3369 reviewed (commented):
The `palaeorotate` function works pretty well and is implemented quite efficiently. Most of my comments are details of suggestions for improvements. Great job!

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
-    #error handling
-    if (!exists("x") | !is.data.frame(x)) {
-      stop("Please supply x as a dataframe")
+palaeorotate <- function(occdf, lng = "lng", lat = "lat", age = "age",
```
I think so too :-D

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
 
-    if (sum(x$lat > 90) != 0 | sum(x$lat < -90) != 0) {
-      stop("Latitude should be more than -90 and less than 90")
+    if (any(!is.numeric(occdf[, lat]) | is.na(occdf[, lat])) ||
```
Not 100% sure what you mean here @Buffan3369 (sorry!). This is basically doing a check to see if any latitudes are NAs or non-numerics. This implementation forces the user to remove any of these data (this way they know exactly what data are being handled).  A user may have entered their data mistakenly as non-numeric (as character strings for example), i.e. "54.5", instead of 54.5, or they may not realise they have NA values. 

In a numeric vector, NAs are interpreted as numerics. For example:

```
vec <- c(5, 2, 4, NA, 5)
is.numeric(vec[4])
is.na(vec[4])
```

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
-    if (sum(x$lat > 90) != 0 | sum(x$lat < -90) != 0) {
-      stop("Latitude should be more than -90 and less than 90")
+    if (any(!is.numeric(occdf[, lat]) | is.na(occdf[, lat])) ||
+        any(!is.numeric(occdf[, lng]) | is.na(occdf[, lng])) ||
```
As above

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
-      stop("Latitude should be more than -90 and less than 90")
+    if (any(!is.numeric(occdf[, lat]) | is.na(occdf[, lat])) ||
+        any(!is.numeric(occdf[, lng]) | is.na(occdf[, lng])) ||
+        any(!is.numeric(occdf[, age]) | is.na(occdf[, age]))) {
```
As above.

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
     }
 
-    #reconstruct coordinates
+    if (sum(occdf[, lat] > 90) != 0 || sum(occdf[, lat] < -90) != 0) {
```
I agree this would work nicely! Updated!

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
 
-    #generate temp directory and download files
+    if (sum(occdf[, lng] > 180) != 0 || sum(occdf[, lng] < -180) != 0) {
+      stop("`lng` values should be >= -180\u00B0 and l<= 180\u00B0")
```
Removed, thanks!

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
 #'
 #' #Rotate the data
-#' palaeorotate(x = x)
+#' palaeorotate(occdf = tetrapods)
```
This is intentionally implemented. The issue here is not with the function but the models/data. It is 'normal' that palaeocoordinates can not be generated for all occurrences because they don't intersect with a plate polygon. If this happens in the grid method and not in the point method, this is because the centroid of the cell the occurrence would be linked to does not fall in the plate polygon.

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
+        c("p_lng_MERDITH2021", "p_lat_MERDITH2021")
+      colnames(scotese) <-
+        c("p_lng_PALEOMAP", "p_lat_PALEOMAP")
+      colnames(wright) <- c("p_lng_WRIGHT2013", "p_lat_WRIGHT2013")
```
Added! Thanks!

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
+        mx <- max(as.numeric(uncertain_lng[i, ]))
+        mn <- min(as.numeric(uncertain_lng[i, ]))
 
         range <- abs((mx %% 360) - (mn %% 360))
```
Given that a lot of the models are coming from the same research groups and their frameworks are being 'recycled', I think SD is misleading. Of course models from the same research group (or using the same base model) are going to be more similar than from different research groups (in my opinion at least).

LewisAJones commented:
> The `palaeorotate` function works pretty well and is implemented quite efficiently. Most of my comments are details of suggestions for improvements. Great job!

Thanks for your comments Lucas! I've implemented those changes or responded where needed!

LewisAJones commented on R/palaeorotate.R:
```diff
@@ -1,246 +1,478 @@
+    }
 
-    #generate temp directory and download files
+    if (sum(occdf[, lng] > 180) != 0 || sum(occdf[, lng] < -180) != 0) {
```
Implemented!

KEichenseer reviewed (approved):
Hi Lewis, 
fantastic job again with this function! I love the API feature :-)

Comments:
- I noticed that the "API" method results in different `p_lng` than the "grid" method with the Meredith model, even if the actual `rot_age`, `rot_lng` and `rot_lat` are specified in the API call. `p_lat` is identical, e.g:
![image](https://user-images.githubusercontent.com/40391500/189475313-c3e8d5b8-6a3e-4cca-8efc-480dd75e4673.png)
This discrepancy seems to exist mainly for the Mesozoic, at age = 300 there is barely any difference.
For the "PALEOMAP" model, p_lng is identical with both methods.
- Add a warning or error if ages are negative? As of now they're being rotated at age 0, or the standard "not able to rotate" warning is produced.
- I suggest to change the documentation for lng, lat, age to say `[...] treated as longitude (defaults to "lng").`, to make it clear that these defaults are already specified. Also, probably specify defaults for the other parameters as well.
- documentation of "Value": Change last sentence to
The "point" method uses the input coordinates and age as the reference; reference coordinates are therefore not returned.
- Reviewers: add Lucas Buffan

Other than that, this is ready! Very cool function.
Kilian

- (Not for now, but maybe an idea for the next palaeoverse release: The processing of multiple API calls could possibly be sped up with something like this: https://www.omegahat.net/RCurl/concurrent.html)

LewisAJones commented:
> Hi Lewis, fantastic job again with this function! I love the API feature :-)

Thanks for all the great comments Kilian, see below for responses/implementations!

> 
> Comments:
> 
> * I noticed that the "API" method results in different `p_lng` than the "grid" method with the Meredith model, even if the actual `rot_age`, `rot_lng` and `rot_lat` are specified in the API call. `p_lat` is identical, e.g:
>   ![image](https://user-images.githubusercontent.com/40391500/189475313-c3e8d5b8-6a3e-4cca-8efc-480dd75e4673.png)
>   This discrepancy seems to exist mainly for the Mesozoic, at age = 300 there is barely any difference.
>   For the "PALEOMAP" model, p_lng is identical with both methods.

This is odd and has me stumped. I've tested a few different ages and it seems to be at 78 Ma the coordinates start to diverge.... but only for p_lng using the Merdith model. All I can think of is that the version of the model used by the GPlates API is slightly different to that from the publication. To ensure that this is consistent, I could try rebuilding the rotation files using the GPlates API... this could take some time but might be worthwhile to ensure consistency between rotation approaches? 

> * Add a warning or error if ages are negative? As of now they're being rotated at age 0, or the standard "not able to rotate" warning is produced.

Added.

> * I suggest to change the documentation for lng, lat, age to say `[...] treated as longitude (defaults to "lng").`, to make it clear that these defaults are already specified. Also, probably specify defaults for the other parameters as well.

Updated.

> * documentation of "Value": Change last sentence to
>   The "point" method uses the input coordinates and age as the reference; reference coordinates are therefore not returned.

Updated.

> * Reviewers: add Lucas Buffan

Added.

> Other than that, this is ready! Very cool function. Kilian

Thanks Kilian!

> * (Not for now, but maybe an idea for the next palaeoverse release: The processing of multiple API calls could possibly be sped up with something like this: https://www.omegahat.net/RCurl/concurrent.html)

Hmmm I will check this out for future updates!

LewisAJones commented:
Hey @KEichenseer, thanks again for your previous review comments. Since your last review, I have gone a bit overboard with my changes. However, I think these have greatly improved the function. To summarise the changes:

- The palaeorotation (reconstruction) grids have now been generated via the GPlates API system so they exactly match. They are also now available for all models that cover the Phanerozoic. I felt it was important to document this process and have a fixed repo for this data. As such, I have made an R package which stores and documents all of this: https://github.com/LewisAJones/palaeorotate

- I have added an additional measure of uncertainty. This measure is the maximum great circle distance between the reconstructed coordinates for each point from all models. 

- I have improved the API feature so that it automatically retries if the API call fails. 

- The API call now splits the data into chunks so it can handle larger datasets.

- I have added a `round` argument so that users can round their data to n decimal places (this can be useful for speeding up the API call by having fewer unique coordinate pairs).

KEichenseer commented:
Hi Lewis, 

great job once again! It all seems to work now :-). And well done for resolving that Merdith grid / point discrepancy.

Not much else to say:

- On the palaeorotate package description, I'd mention the file size of the palaeorotate package, and a warning that installation takes a long time.

- The warning "defined `lng`, `lat`, and `age` not found in `occdf`" could be changed to "defined `lng`, `lat`, OR `age` not found in `occdf`" (I had the case where only age was not defined and I was confused as to why it hadn't fount the other columns).

Kilian

LewisAJones commented:
> Hi Lewis,
> 
> great job once again! It all seems to work now :-). And well done for resolving that Merdith grid / point discrepancy, and for improving the API calls.
> 
> Not much else to say:
> 
> * On the palaeorotate package description, I'd mention the file size of the palaeorotate package, and a warning that installation takes a long time.
> * The warning "defined `lng`, `lat`, **and** `age` not found in `occdf`" could be changed to "defined `lng`, `lat`, **or** `age` not found in `occdf`" (I had the case where only `age` was not defined and I was confused as to why it hadn't fount the other columns).
> 
> Kilian

Hi Kilian,

Thanks for taking a final look at this! I think this function is finally ready to be integrated, so will go ahead and merge.

The package isn't actually being installed, it is just downloading the data. However, I've added a note on file size. 

I've updated the warning as requested, thanks for the heads up!

Lewis
